File: apps/ai/app/domain/agent_configuration.py
"""Tenant-owned AI agent runtime configuration. No FastAPI or provider SDKs."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Mapping

from app.domain.orchestration import ModelRoute, SAFE_FALLBACK_REPLY
from app.domain.tools import TOOL_NAMES, ToolName, parse_tool_name

FallbackMode = Literal["provider_then_heuristic", "canned_reply", "handoff"]
CitationPolicy = Literal["required", "preferred", "off"]

FALLBACK_MODES: tuple[FallbackMode, ...] = (
    "provider_then_heuristic",
    "canned_reply",
    "handoff",
)
CITATION_POLICIES: tuple[CitationPolicy, ...] = ("required", "preferred", "off")
HANDOFF_REPLY = "I am connecting you with a teammate who can help with this."
UNKNOWN_REPLY = (
    "I do not have enough information in the knowledge base to answer that. "
    "I can connect you with a teammate."
)


@dataclass(frozen=True, slots=True)
class AgentRuntimeConfig:
    model: str | None = None
    quality_model: str | None = None
    temperature: float | None = None
    max_output_tokens: int | None = None
    max_input_tokens: int | None = None
    system_prompt: str = ""
    enabled_tools: tuple[ToolName, ...] | None = None
    fallback_mode: FallbackMode = "provider_then_heuristic"
    fallback_reply: str | None = None
    fallback_max_retries: int | None = None
    citation_policy: CitationPolicy = "preferred"
    refuse_unknown: bool = True
    refuse_off_topic: bool = True
    language_lock: bool = True
    redact_pii: bool = False

    def merged_instructions(self, instructions: str) -> str:
        extra = self.system_prompt.strip()
        if not extra:
            return instructions
        return f"{instructions}\n\nOperator prompt:\n{extra}"

    def policy_instructions(self) -> str:
        parts: list[str] = []
        if self.citation_policy == "required":
            parts.append(
                "Only answer from the knowledge excerpts. If they do not contain the answer, say you are unsure."
            )
        elif self.citation_policy == "off":
            parts.append("Do not mention sources or citations in the customer-facing reply.")
        else:
            parts.append("Prefer knowledge excerpts when they are relevant.")
        if self.refuse_unknown:
            parts.append("Do not invent policies, prices, order numbers, or account data.")
        if self.refuse_off_topic:
            parts.append("Decline off-topic requests and offer a human handoff.")
        if self.language_lock:
            parts.append("Reply only in the configured language.")
        if self.redact_pii:
            parts.append("Do not echo emails, phone numbers, or other personal identifiers.")
        return " ".join(parts)

    def apply_route(self, route: ModelRoute) -> ModelRoute:
        model = route.model
        if route.name == "quality":
            chosen = (self.quality_model or self.model or "").strip()
            if chosen:
                model = chosen
        else:
            chosen = (self.model or "").strip()
            if chosen:
                model = chosen
        temperature = self.temperature if self.temperature is not None else route.temperature
        return ModelRoute(
            name=route.name,
            model=model,
            temperature=temperature,
            retrieve=route.retrieve,
            json_mode=route.json_mode,
        )

    def safe_reply(self, *, reason: str) -> str:
        del reason
        if self.fallback_mode == "handoff":
            return HANDOFF_REPLY
        canned = (self.fallback_reply or "").strip()
        if canned:
            return canned
        return SAFE_FALLBACK_REPLY


def parse_agent_runtime_config(raw: Mapping[str, object] | None) -> AgentRuntimeConfig | None:
    if raw is None:
        return None
    tools_raw = raw.get("enabledTools")
    enabled: tuple[ToolName, ...] | None = None
    if isinstance(tools_raw, list):
        enabled = tuple(parse_tool_name(str(item)) for item in tools_raw)
    elif tools_raw is None:
        enabled = TOOL_NAMES
    fallback_mode = raw.get("fallbackMode")
    citation_policy = raw.get("citationPolicy")
    return AgentRuntimeConfig(
        model=_optional_str(raw.get("model")),
        quality_model=_optional_str(raw.get("qualityModel")),
        temperature=_optional_float(raw.get("temperature")),
        max_output_tokens=_optional_int(raw.get("maxOutputTokens")),
        max_input_tokens=_optional_int(raw.get("maxInputTokens")),
        system_prompt=str(raw.get("systemPrompt") or ""),
        enabled_tools=enabled,
        fallback_mode=fallback_mode if fallback_mode in FALLBACK_MODES else "provider_then_heuristic",
        fallback_reply=_optional_str(raw.get("fallbackReply")),
        fallback_max_retries=_optional_int(raw.get("fallbackMaxRetries")),
        citation_policy=citation_policy if citation_policy in CITATION_POLICIES else "preferred",
        refuse_unknown=_optional_bool(raw.get("refuseUnknown"), True),
        refuse_off_topic=_optional_bool(raw.get("refuseOffTopic"), True),
        language_lock=_optional_bool(raw.get("languageLock"), True),
        redact_pii=_optional_bool(raw.get("redactPii"), False),
    )


def _optional_str(value: object) -> str | None:
    if not isinstance(value, str):
        return None
    trimmed = value.strip()
    return trimmed or None


def _optional_float(value: object) -> float | None:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    number = float(value)
    if number < 0 or number > 2:
        return None
    return number


def _optional_int(value: object) -> int | None:
    if isinstance(value, bool) or not isinstance(value, int):
        return None
    return value


def _optional_bool(value: object, default: bool) -> bool:
    if isinstance(value, bool):
        return value
    return default

File: packages/contracts/src/agent-configuration.ts
import { TOOL_NAMES, type ToolName, isToolName } from './tools.js';

export const AI_AGENT_CONFIGURATION_SCHEMA_VERSION = 1 as const;

export const AI_AGENT_MODELS = ['gpt-4o-mini', 'gpt-4o', 'gpt-4.1-mini', 'gpt-4.1'] as const;
export type AiAgentModelId = (typeof AI_AGENT_MODELS)[number];

export const AI_AGENT_FALLBACK_MODES = ['provider_then_heuristic', 'canned_reply', 'handoff'] as const;
export type AiAgentFallbackMode = (typeof AI_AGENT_FALLBACK_MODES)[number];

export const AI_AGENT_CITATION_POLICIES = ['required', 'preferred', 'off'] as const;
export type AiAgentCitationPolicy = (typeof AI_AGENT_CITATION_POLICIES)[number];

export const DEFAULT_AI_AGENT_ENABLED_TOOLS: readonly ToolName[] = TOOL_NAMES;

export type AiAgentRuntimeConfigDto = {
  readonly schemaVersion: typeof AI_AGENT_CONFIGURATION_SCHEMA_VERSION;
  readonly model: AiAgentModelId;
  readonly qualityModel: AiAgentModelId;
  readonly temperature: number;
  readonly maxOutputTokens: number;
  readonly maxInputTokens: number;
  readonly systemPrompt: string;
  readonly enabledTools: readonly ToolName[];
  readonly fallbackMode: AiAgentFallbackMode;
  readonly fallbackReply: string | null;
  readonly fallbackMaxRetries: number;
  readonly citationPolicy: AiAgentCitationPolicy;
  readonly refuseUnknown: boolean;
  readonly refuseOffTopic: boolean;
  readonly languageLock: boolean;
  readonly redactPii: boolean;
};

export type AiAgentConfigurationDto = AiAgentRuntimeConfigDto & {
  readonly id: string;
  readonly organizationId: string;
  readonly createdAt: string;
  readonly updatedAt: string;
};

export type UpdateAiAgentConfigurationRequest = {
  readonly model?: AiAgentModelId;
  readonly qualityModel?: AiAgentModelId;
  readonly temperature?: number;
  readonly maxOutputTokens?: number;
  readonly maxInputTokens?: number;
  readonly systemPrompt?: string;
  readonly enabledTools?: readonly ToolName[];
  readonly fallbackMode?: AiAgentFallbackMode;
  readonly fallbackReply?: string | null;
  readonly fallbackMaxRetries?: number;
  readonly citationPolicy?: AiAgentCitationPolicy;
  readonly refuseUnknown?: boolean;
  readonly refuseOffTopic?: boolean;
  readonly languageLock?: boolean;
  readonly redactPii?: boolean;
};

export type AiAgentConfigurationCatalogDto = {
  readonly models: readonly AiAgentModelId[];
  readonly tools: readonly ToolName[];
  readonly fallbackModes: readonly AiAgentFallbackMode[];
  readonly citationPolicies: readonly AiAgentCitationPolicy[];
};

export type AiAgentConfigurationResponse = {
  readonly configuration: AiAgentConfigurationDto;
  readonly catalog: AiAgentConfigurationCatalogDto;
};

export function isAiAgentModelId(value: unknown): value is AiAgentModelId {
  return typeof value === 'string' && (AI_AGENT_MODELS as readonly string[]).includes(value);
}

export function isAiAgentFallbackMode(value: unknown): value is AiAgentFallbackMode {
  return typeof value === 'string' && (AI_AGENT_FALLBACK_MODES as readonly string[]).includes(value);
}

export function isAiAgentCitationPolicy(value: unknown): value is AiAgentCitationPolicy {
  return typeof value === 'string' && (AI_AGENT_CITATION_POLICIES as readonly string[]).includes(value);
}

export function isAiAgentRuntimeConfigDto(value: unknown): value is AiAgentRuntimeConfigDto {
  if (!isRecord(value)) {
    return false;
  }

  return (
    value.schemaVersion === AI_AGENT_CONFIGURATION_SCHEMA_VERSION &&
    isAiAgentModelId(value.model) &&
    isAiAgentModelId(value.qualityModel) &&
    isTemperature(value.temperature) &&
    isOutputTokens(value.maxOutputTokens) &&
    isInputTokens(value.maxInputTokens) &&
    typeof value.systemPrompt === 'string' &&
    isToolNameArray(value.enabledTools) &&
    isAiAgentFallbackMode(value.fallbackMode) &&
    isNullableString(value.fallbackReply) &&
    isRetryCount(value.fallbackMaxRetries) &&
    isAiAgentCitationPolicy(value.citationPolicy) &&
    typeof value.refuseUnknown === 'boolean' &&
    typeof value.refuseOffTopic === 'boolean' &&
    typeof value.languageLock === 'boolean' &&
    typeof value.redactPii === 'boolean'
  );
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isNullableString(value: unknown): value is string | null {
  return value === null || typeof value === 'string';
}

function isToolNameArray(value: unknown): value is ToolName[] {
  return Array.isArray(value) && value.every(isToolName);
}

function isTemperature(value: unknown): value is number {
  return typeof value === 'number' && Number.isFinite(value) && value >= 0 && value <= 2;
}

function isOutputTokens(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 64 && value <= 4096;
}

function isInputTokens(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 512 && value <= 32_000;
}

function isRetryCount(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 1 && value <= 5;
}
